test/unit/controller/make_a_demo_payment_controller/confirm_controller_test.js

```diff
@@ -0,0 +1,200 @@
+'use strict'
+
+// NPM dependencies
+const supertest = require('supertest')
+const {expect} = require('chai')
+const cheerio = require('cheerio')
+const lodash = require('lodash')
+const nock = require('nock')
+
+// Local dependencies
+const {getApp} = require('../../../../server')
+const {getMockSession, createAppWithSession, getUser} = require('../../../test_helpers/mock_session')
+const paths = require('../../../../app/paths')
+const {randomUuid} = require('../../../../app/utils/random')
+const {validCreateProductRequest, validCreateProductResponse} = require('../../../fixtures/product_fixtures')
+
+const {PUBLIC_AUTH_URL, PRODUCTS_URL} = process.env
+const GATEWAY_ACCOUNT_ID = 929
+const PAYMENT_DESCRIPTION = 'Pay your window tax'
+const PAYMENT_AMOUNT = '20.00'
+const VALID_USER = getUser({
+  gateway_account_ids: [GATEWAY_ACCOUNT_ID],
+  permissions: [{name: 'transactions:read'}]
+})
+const VALID_CREATE_TOKEN_REQUEST = {
+  account_id: GATEWAY_ACCOUNT_ID,
+  created_by: VALID_USER.email,
+  description: `Token for Demo Payment`
+}
+const VALID_CREATE_TOKEN_RESPONSE = {token: randomUuid()}
+const VALID_CREATE_PRODUCT_REQUEST = validCreateProductRequest({
+  name: PAYMENT_DESCRIPTION,
+  payApiToken: VALID_CREATE_TOKEN_RESPONSE.token,
+  price: PAYMENT_AMOUNT * 100,
+  gatewayAccountId: GATEWAY_ACCOUNT_ID
+}).getPlain()
+const VALID_CREATE_PRODUCT_RESPONSE = validCreateProductResponse(VALID_CREATE_PRODUCT_REQUEST).getPlain()
+
+describe.only('make a demo payment - confirm controller', () => {
+  describe(`when both paymentDescription and paymentAmount exist in the session`, () => {
+    describe(`when the API token is successfully created`, () => {
+      describe(`and the product is successfully created`, () => {
+        let result, $, session
+        before(done => {
+          nock(PUBLIC_AUTH_URL).post('', VALID_CREATE_TOKEN_REQUEST).reply(201, VALID_CREATE_TOKEN_RESPONSE)
+          nock(PRODUCTS_URL).post('/v1/api/products', VALID_CREATE_PRODUCT_REQUEST).reply(201, VALID_CREATE_PRODUCT_RESPONSE)
+          session = getMockSession(VALID_USER)
+          lodash.set(session, 'pageData.makeADemoPayment.paymentDescription', PAYMENT_DESCRIPTION)
+          lodash.set(session, 'pageData.makeADemoPayment.paymentAmount', PAYMENT_AMOUNT)
+          supertest(createAppWithSession(getApp(), session))
```

Comment:
I can see why you've put this in the before() so that you can have separate tests making assertions about different aspects of the expected behaviour in this circumstance.

However, I would like if at all possible to follow the [ArrangeActAssert](http://wiki.c2.com/?ArrangeActAssert) pattern for tests, and in particular to avoid doing the 'act' step in `before` blocks. IMHO both "act" and "assert" should occur explicitly in each test. 

I would consider `createAppWithSession()` to be part of "arrange" but `get()` is "act".

Reply:
I like this definition of Arrange, Act, Assert I'm not sure I agree with both "act" and "assert" being in each test (`it` block):
- putting the "act" in each `it` block would duplicate an awful lot code if you keep the separate tests for different aspects of the expected behaviour
- if you consolidate the `it` blocks, you end up with poor test feedback when you make future changes, as each `it` block ends up testing too much and you only see what assertion (i.e. `expected: true, actual: false`) has failed not what particular expected behaviour has failed to occur
- by having multiple asynchronous `it` blocks rather than handling all asynchronous code in a `before` block your tests can end up taking longer (particularly when doing something slower like pact testing)

If we really wanted to maintain separation between "arrange" and "act" an option could be to do two`before` blocks to effectively separate this code (arrange generally  is synchronous anyway). Effectively giving us:
`before(() => {...}) // Arrange`
`before(done => {...}) // Act`
`it(() => {...}) // Assert`

Mocha calls before blocks in order so this would work in practice.

Reply:
the aim is not to have each of arrange/act/assert is in a separate `it` block, but rather to have a pattern which clearly separates each of those things. In the c2 wiki they talk about having segments separated by a blank line. That would be sufficient. 

The important thing is that:
a) there's a consistent pattern used in all tests
b) the three parts are separated in a clear way

I do have a preference for having the 'act' and 'assert' parts within the `it` block. The reason for that preference is that when I'm reading a test, the first thing that I look for is *what* is being tested, and in order to find that I'm looking for the 'act' part. So I expect that to be clearly visible in each test step. 

Regarding the risk of duplication/repetition, that's absolutely fine. 

Test code should be [DAMP not DRY](https://stackoverflow.com/a/11837973).